sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala

```diff
@@ -145,6 +145,11 @@ abstract class PrunedFilteredScan extends BaseRelation {
  * for experimentation.
  */
 @Experimental
-abstract class CatalystScan extends BaseRelation {
+trait CatalystScan extends BaseRelation {
   def buildScan(requiredColumns: Seq[Attribute], filters: Seq[Expression]): RDD[Row]
 }
+
+@DeveloperApi
+trait InsertableRelation extends BaseRelation {
+  def insertInto(data: DataFrame, overwrite: Boolean): Unit
```

Comment:
The method name looks a bit weird when you put them together: `relation.insertInto(dataFrame)`, seems that we are inserting a relation into a data frame... Maybe just `insert`?